src/Native/System.Native/pal_networking.cpp

```diff
@@ -428,7 +429,168 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR
+#if !HAVE_GETHOSTBYNAME_R
+static int copy_hostent(struct hostent* from, struct hostent* to,
+                        char* buffer, size_t buflen)
+{
+    int i, len;
+    char* ptr = buffer;
+
+    to->h_addrtype = from->h_addrtype;
+    to->h_length = from->h_length;
+    to->h_name = static_cast<char*>(ptr);
+
+    /* copy hostname to buffer */
+    len = static_cast<int>(strlen(from->h_name)+1);
+    SafeStringCopy(ptr, buflen, from->h_name);
+    ptr += len;
+
+    if (strlen(ptr)+1 > buflen)
+    {
+        return ERANGE;
+    }
+
+    /* copy aliases to buffer */
+    to->h_aliases = reinterpret_cast<char**>(ptr);
+    for(i = 0; from->h_aliases[i]; i++);
+    ptr += (i+1) * static_cast<int>(sizeof(char*));
+
+    if (strlen(ptr)+1 > buflen)
+    {
+        return ERANGE;
+    }
+
+    for (i = 0; from->h_aliases[i]; i++)
+    {
+        len = static_cast<int>(strlen(from->h_aliases[i])+1);
+        if (len >= 0 && static_cast<size_t>((ptr-buffer)+len) < buflen)
+        {
+            to->h_aliases[i] = ptr;
+            SafeStringCopy(ptr, static_cast<int>(ptr-buffer), from->h_aliases[i]);
+            ptr += len;
+        }
+        else
+        {
+            return ERANGE;
+        }
+    }
+
+    if (strlen(to->h_aliases[i])+1 > buflen)
```

Comment:
I don't understand this code. 1) i shouldn't even be in range for h_aliases at this point, as we just finished walking to the point where from->h_aliases[i] is null. 2) The lengut of that one alias shouldn't be what matters, but rather how much of the buffer we've used thus far. 3) We know exactly where we are in the buffer, we don't need to be calling functions like strlen that need to recompute it.